docs/design/uptime-reputation-calculation.md

```diff
@@ -0,0 +1,122 @@
+
+
+## Rationale
+
+The approach described in this document present the following advantages:
+
+1. It's clear why an _SN_ get disqualified due to failed uptime checks.
+1. It has a clear proof for an _SNOs_ why their _SNs_ get disqualified due to failed uptime check in case that they request it before the designed time to clean up the historical failed uptime checks data.
+1. It allows to report to _SN_ at any how far is to be disqualified by failing uptime checks.
+
+And the following disadvantages
+
+1. It requires to store, for some period of time, historical for nodes which have failed at least one uptime check.
+1. It requires more resources than the current implemented approach.
+1. It doesn't calculate the exact time that an _SN_ is offline; once an _SN_ has a failed uptime check the time elapsed between the last failed uptime check and the next succeeded uptime check causes not to have the accurate offline time.
+1. The _"score"_ obtained from uptime check failures, without any further modification, doesn't integrate with the current _SN reputation system_.
+
+While one of our goals is to reduce the number of resources used by the satellite we see that the advantages beat the disadvantages considering that the current system got disabled after some nodes got disqualified unfairly and/or unclear.
+
+On the other hand, we think that the number of resources required by this new system shouldn't be too much and we believe that we can have an implementation that scales up.
+
+## Implementation
+
+### DB
+
+A database is required to store information about _SNs_ uptime check failures. This section presents the new required database schema and a brief explanation what it will store.
+
+Only one database table is required and its schema is the following (pseudo code):
+
+```sql
+CREATE TABLE failed_uptime_checks (
+    node_id       BYTEA NOT NULL,
+    when          TIMESTAMP WITH TIMEZONE NOT NULL, -- When the SN failed the an uptime check until it succeeds a future uptime check
+    back_online   TIMESTAMP WITH TIMEZONE, -- When SN has succeeded an uptime check after the last one failed
+    disqualified  TIMESTAMP WITH TIMEZONE  -- When the SN has been disqualified by any disqualification system (uptimes, audits, etc.)
+)
+```
+
+The table will be filled with _SNs_ which have failed at least one uptime check during the ongoing month.
+
+Each row in the table indicates when an _SN_ failed the first uptime check (`when`), while the _SN_ not succeed any future uptime check (`back_online` is `NULL`); when the _SN_ is back online any future uptime check failure will insert a new row to mark a new period of time that the node is offline again.
+
+When the _SN_ is disqualified by any disqualification system and the _SN_ isn't back online, the `disqualified` column of the row will be updated for not receiving more uptime checks because we don't do business with disqualified _SNs_.
```

Comment:
Could we gain some efficiency by removing them from this table completely instead?  Or do we think that won't be worth the effort?

Reply:
:thinking: we actually could.

I will present that in today's meeting.

Reply:
This also will be discussed in the next meeting that I'll convoke.